Don't show error on daemon side when inspect image

hqhq opened the pull request:
Fixes: #10773

Container or image not found should not be daemon's fault, so
we can return error to client, but don't show error on daemon side.

This commit introduce an errtype package, this could be the first
step we standard err types in Docker in necessary places.

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

runcom commented:
What's reported in #10773 is somehow wrong to me, let me explain. `docker inspect` is cli side and it uses apis. From a pure clinet-api prospective that error showed in daemon logs is totally correct meaning if you have requested a non existent resource it's correct to return a 404 error. So if the client (docker cli) makes a call to container/json and it fails with a not found error then api have to log this as a 404 error. 
Problem here, to me, resides in how `docker inspect` works and I'm sure a better fix for this is to modify the command to query only images or containers endpoint. It's not really correct to shadow an api error only because the client is messing around with endpoint calls. My 2 cents..

hqhq committed:
Don't show error on daemon side when inspect image

Fixes: #10773

Container or image not found should not be daemon's fault, so
we can return error to client, but don't show error on daemon side.

This commit introduce an errtype package, this could be the first
step we standard err types in Docker in necessary places.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

duglin commented:
See #9509  its kind of related.

But the bigger issue is Docker's logging/error reporting in general.
Just for comparison purposes I looked at what my Tomcat does and what I observed is:
- when there's a user error (e.g. 404) I don't see anything in the logs
- when there's an error in the webapp (e.g. bad jsp) then I see an error in the logs

I think the rule here is that "user errors are normal and therefore nothing worth noting in the daemon logs".   By normal, I mean I expect clients to send bad data and as long as we handle it properly then the daemon acted as it should so there's no reason to make a note of it in the logs.  Note, this isn't the same thing as some kind of audit system that may want to track all traffic, I'm just talking about the portion of the daemon log file which we use for debugging, and tracking of things that we really should do something about.

So, note that Docker's log actually includes 3 things:
- the auditing info I mentioned
- user error info
- daemon error info - things we should investigate/fix - e.g. bad state in the DB

Whether audit info and daemon error info should be split or not is questionable, but I don't see a point in  fulling up our logs with user error info.  We track the incoming request (might be nice to include the response status code - e.g. 404 - like tomcat does) but that's it. I don't think we should generate an ERROR in there because to me ERRORs are things that we, as docker devs, should look at and try to fix.  Ideally, we should never see anything in the logs except the auditing info.  Anything else should be a cause for concern.

Just my 2 cents.

duglin commented:
Hit send too soon, darn.... this may mean we need to modify our error processing flow to distinguish between a user error and a daemon error.  However, once we have #9509 we may get it implicitly because each error will have an http status code associated with it and if its >= 500 then its a real daemon issue to be looked at, while < 500 is more like a user-error.  I think. Would need to look at each error case to know for sure - but at least that's a consistent model to follow.

runcom commented:
@duglin ok agree with what you said (hope to get this soon) but I think downgrading that 404 error to info is not correct here to fix #10773, better approach is #13187

duglin commented:
From https://github.com/docker/docker/issues/10773#issuecomment-106330299:
I think we should close this issue, not because the premise is wrong but rather because this is just the way docker works right now. For better or worse, user caused errors are logged as ERRORs in the daemon's log. I know this isn't technically a user error, but from the daemon's perspective it is.
Once we get #9509 then we can "fix" it so that the only ERRORs that appear in the daemon logs are things that are really wrong with the daemon and not user cause issues.

Plus, I don't think docker inspect is special w.r.t. this issue and we should solve it globally not just for this one API.

thaJeztah commented:
Am I right that the consensus here is that this is *not* the desired approach? Or does this need input from other maintainers?

hqhq commented:
@thaJeztah As I said in https://github.com/docker/docker/issues/10773#issuecomment-106648178, I think we haven't had any consensus now, maybe we need input from other maintainers, like @cpuguy83 who suggested low level the error entry.

icecrime commented:
Yeah sorry @hqhq I think I'm -1 on this solution too =/
